operations.ts

```diff
@@ -0,0 +1,138 @@
+import { Type, UNKNOWN } from './value';
+
+/**
+ * The expected return type of a typical binary math expression, given
+ * its input types.
+ *
+ * @param a The type of the first argument.
+ * @param b The type of the second argument.
+ *
+ * @returns The expected type of the operation.
+ */
+export function binaryMathReturnType(a: Type, b: Type): Type | typeof UNKNOWN {
+  // TODO: This is an extremely error-prone approach to this problem.
+  switch (a) {
```

Comment:
I think that, instead of trying to enumerate every combination, we can do something like this:

1. Find the highest precedence type from "boolean", "integer", "float". For example, between "boolean" and "float", the highest precedence is "float". That's the return type of the operation.
2. Check "wouldCast" for the other argument into that precedence. If it fails, return Type.Failure instead of the highest precedence type.

For math operations, cases of "anything else" will always fail. For addition, we can add "string" into the mix but maintain the same algorithm.

I'm also starting to think that I need to flip this around and implement in the runtime before I implement in the compiler. Think about how it worked with `cast`/`wouldCast` - I implemented an actual cast and tested it, then wrote `wouldCast` and extended the tests to cross-reference the result. This will also pay higher dividends early - it will immediately implement the correct semantics in the runtime, instead of attempting to introduce type safety on top of a shaky foundation.